src/journal/journal.c

```diff
@@ -0,0 +1,678 @@
+{
+	if (telem_journal) {
+                free(telem_journal->boot_id);
+                free(telem_journal->journal_file);
+                fclose(telem_journal->fptr);
+                free(telem_journal);
+        }
+}
+
+/* Exported function */
+int print_journal(TelemJournal *telem_journal, char *classification,
+                   char *record_id, char *event_id, char *boot_id)
+{
+        int n = 0;
+        int rc = 0;
+        int count = 0;
+        char str_time[80] = {0};
+        char *line = NULL;
+        size_t len = 0;
+        FILE *journal_fileptr = NULL;
+        struct tm ts;
+        struct JournalEntry *entry = NULL;
+
+        if (telem_journal == NULL) {
+                return -1;
+        }
+
+        journal_fileptr = fopen(telem_journal->journal_file, "r");
+        if (!journal_fileptr) {
+                return -1;
+        }
+
+        // Advance file pointer to line n
+        n = telem_journal->record_count - telem_journal->record_count_limit;
+        if (n > 0) {
+                rc = skip_n_lines(n, journal_fileptr);
+                if (rc == -1) {
+                        return rc;
+                } else if (rc != 0) {
+                        fprintf(stderr, "Error: ", strerror(rc));
+                }
+        }
+
+        while (getline(&line, &len, journal_fileptr) != -1) {
+                deserialize_journal_entry(line, &entry);
+                if (entry) {
+                        /* filter entry out if one is provided */
+                        if (record_id != NULL && strncmp(entry->record_id, record_id, ID_LEN) != 0) {
+                                continue;
+                        }
+                        if (classification != NULL && strncmp(entry->classification, classification, strlen(entry->classification)) != 0) {
+                                continue;
+                        }
+                        if (boot_id != NULL && strncmp(entry->boot_id, boot_id, BOOTID_LEN) != 0) {
+                                continue;
+                        }
+                        if (event_id != NULL && strncmp(entry->event_id, event_id, strlen(entry->event_id)) != 0) {
+                                continue;
+                        }
+                        /* end filters section */
+                        ts = *localtime(&entry->timestamp);
+                        if (strftime(str_time, sizeof(str_time), "%a %Y-%m-%d %H:%M:%S %Z", &ts) == 0) {
+                                continue;
+                        }
+
+                        fprintf(stdout, "%-30s %s %s %s %s\n", entry->classification, str_time, entry->record_id, entry->event_id, entry->boot_id);
+                        count++;
+                }
+                free_journal_entry(entry);
+        }
+        free(line);
+        fclose(journal_fileptr);
+
+        return count;
+}
+
+/**
+ * Validate classification
+ * TODO: move to util or common and use it in telemetry.c
+ *
```

Comment:
I would like to see a comment here saying what a valid classification is, so one can check the code agrees with it.